metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -2035,27 +2035,16 @@ void BOSS::call_paths(Call<std::vector<edge_index>&&,
             assert(edge);
         } while (edge != start);
 
-        // If |kmers_in_single_form| = true, the edge mask |fetched| is
-        // used in call_path to prevent calling k-mers multiple times.
-        // Otherwise, that check has to be done here, so we check the cycle's
-        // representative node to see if the cycle has been called already.
-        if (!kmers_in_single_form) {
-            edge_index rep = *std::min_element(path.begin(), path.end());
-            if (fetch_and_set_bit(visited.data(), rep, async))
-                return;
-
-            ++progress_bar;
-        }
-
-        for (edge_index edge : path) {
-            if (!fetch_and_set_bit(visited.data(), edge, async))
-                ++progress_bar;
+        // Ensures that call_path is called only once for each cycle
+        edge_index rep = *std::min_element(path.begin(), path.end());
+        if (!fetch_bit(visited.data(), rep, async)) {
+            // TODO: avoid get_node_seq(rep), and maybe pass the whole path
+            ::mtg::graph::boss::call_paths(
```

Comment:
why `call_paths`? at this point, all non-cyclic parts of the graph should have been covered

Reply:
To make it the processing the same as for other unitigs, so there are no special cases. What if not `call_caths`?

Reply:
it's a little bit overkill because of the extra incoming/outgoing checking

Reply:
`process_cycle` already does the same traversal as `call_paths` if you assume that the edge you're on is part of a cycle, the only difference is the order in which they set bits in `visited`. Before, we used to first set the bit at the representative, then set each bit in the path order.

To make it the same as with other unitigs, we'd have to first rotate the cycle so that it starts from the representative edge, then set bits in `visited` in order.

Reply:
This way, the order in which bits are set in `visited` will correspond to paths in the graph for cycles as well.

Reply:
But I start the traversal from the minimizer, so it is indeed the same as any other unitig/contig.

The first thing `call_paths` does is marks that minimizer as visited, so no cycle is called twice as a forward path, and all assumptions are satisfied.

I see, however, that this is overkill, especially taking into account that the entire cycle has already been traversed, so there is no need to do it again. But I just wanted to use a routine that already exists. I guess we need to write another simplified version of `call_paths` adjusted for this case: use the existing path without traversing the graph, but with checks for `visited`.

Reply:
or just iterate through in that order without rotating the path

Reply:
maybe it'll be enough to do something like
```c++
auto rep = std::min_element(path.begin(), path.end());
std::rotate(path.begin(), rep, path.end());
std::rotate(sequence.begin(), sequence.begin() + (rep - path.begin()), sequence.end());
auto begin = path.begin();
while (rep != path.end()) {
    if (fetch_and_set_bit(visited.data(), *rep, async)) {
        if (begin != rep) {
            // do stuff, then call_path
        }
        begin = rep;
    }
    ++rep;
}
```

Reply:
it's not enough, we also need to start traversal for `dual_endpoints` and so on

Reply:
I pushed a patch, but not sure this worth it. Maybe you want to benchmark...?

Reply:
One thing to note is that by the time we get to `process_cycles`, in both cases, all of the remaining k-mers are present in cycles. So when checking cycles in parallel, there are three cases to cover:

1. non-primary mode or none of the cycle's dual k-mers are in the graph:
mark the minimizer as visited, then mark the rest and call

2. all of the cycle's dual k-mers are in the graph:
mark the minimum of the path's and the dual path's minimizers as visited, then mark the rest in both strands and call only the forward path

3. some cycles with partial overlaps (may also overlap itself)
lock, rotate, then go through the path marking both the canonical and non-canonical forms of its k-mers. Use the result of fetch_and_set_bit on the canonical k-mer to decide where to split the cycle for output

I implemented this strategy in a76cfd288b8d7519cdc7d1360dc6633a245eaa53 but I haven't benchmarked it on any large data sets yet.

Reply:
I'll benchmark both approaches and let you know how it goes.